src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,443 @@
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidParameterSpecException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class, algorithm,
+                                                        algParameterSpec, provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider " + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm, algParameterSpec);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new NoSuchAlgorithmException("Algorithm " + algorithm + " not available", nsae);
+        }
+    }
+
+    /**
+     * Derive a key, returned as a {@code Key}.
+     * <p>
+     * TODO: additional description
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code Key} object
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the KDF output and according
+     * to the derivation parameters.
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is invalid or incorrect
+     *     for the type of key to be derived, or specifies a type of output that is not a key (e.g.
+     *     raw data)
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to {@code deriveKey}
+     *     or if all {@code KDFParameterSpec} objects provided at initialization have been
+     *     processed.
+     * @throws InvalidAlgorithmParameterException
+     *     TODO: fill this in
+     */
+    public SecretKey deriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException, InvalidAlgorithmParameterException {
+        return spi.engineDeriveKey(alg, kdfParameterSpec);
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     * <p>
+     * TODO: additional description
+     *
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a byte array whose length matches the length field in the processed
+     * {@code DerivationParameterSpec} and containing the next bytes of output from the key
+     * derivation function.
+     *
+     * @throws InvalidParameterSpecException
+     *     if the {@code DerivationParameterSpec} being applied to this method is of a type other
+     *     than "data".
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to
+     *     {@code deriveData } or if all {@code DerivationParameterSpec} objects have been
+     *     processed.
```

Comment:
I don't think these methods should throw `IllegalStateException` - that conflicts with the thread-safe design of the API - perhaps this was leftover from an earlier iteration?